### PR TITLE
Allow to set Shared Memory size (shmSize) for GenericContainer

### LINKED ISSUE #952: Can't set the shm_size for oracle xe
My oracle xe container is currently not starting because of a lack of shared memory:
"A database XE container needs at least 1 GB of shared memory (/dev/shm)."

My problem here: How can I set the shm_size (shmSize) for OracleContainer respectively GenericContainer?

I've already tried: .withCreateContainerCmdModifier(cmd -> cmd.withMemory((long) 4 * 1024 * 1024)). 

Also tried some settings within the testcontainers.properties. Unfortunately I've not found what params are available and how to set it here.

Any help welcome. Thanks.

### PR DESCRIPTION
**Pull request by fzakaria**
Add the necessary code so that `shmSize` can be set on a GenericContainer.

This is especially useful for when starting Oracle docker images because they need at least 1GB of shared memory space.

> [docker run ... --shm-size=1g oracle/database:11.2.0.2-xe](https://github.com/oracle/docker-images/blob/master/OracleDatabase/SingleInstance/FAQ.md)

Fixes #952

### CODE REVIEW

**Comment by rnorth** [comment]
Thanks @fzakaria.

Looks like CircleCI had an unrelated glitch downloading the gradle wrapper so I've restarted the build.

**Comment by fzakaria** [comment]
Thanks @rnorth  -- looks to have been resolved.

File: core/src/test/java/org/testcontainers/junit/GenericContainerRuleTest.java [hunk 1]
```diff
@@ -124,6 +126,13 @@ public static void setupContent() throws FileNotFoundException {
             .withExtraHost("somehost", "192.168.1.10")
             .withCommand("/bin/sh", "-c", "while true; do cat /etc/hosts | nc -l -p 80; done");
 
+    /**
+     * Redis with shared memory
+     */
+    @ClassRule
+    public static GenericContainer redisWithSharedMemory = new GenericContainer("redis:3.0.2")
```
[the remark this answers is not in the document]
Comment (fzakaria) [reply core/src/test/java/org/testcontainers/junit/GenericContainerRuleTest.java]: @rnorth this is following the same convention in the rest of the document -- it's not clear why Codacy is complaining.

**Comment by fzakaria** [comment]
I updated the PR to have _shmSize_ be simply bytes instead of 'megabytes' to avoid mebibyte vs megabyte confusion.

**Comment by fzakaria** [comment]
Looks like travis-ci had a flaky failure pulling an artifact -- could someone restart.
I believe this PR is ready to merge; thanks.

File: core/src/main/java/org/testcontainers/containers/GenericContainer.java [hunk 2]
```diff
@@ -21,6 +22,7 @@
 import org.apache.commons.compress.archivers.tar.TarArchiveInputStream;
 import org.apache.commons.compress.archivers.tar.TarArchiveOutputStream;
 import org.apache.commons.compress.utils.IOUtils;
+import org.apache.commons.io.FileUtils;
```
Comment (kiview) [inline core/src/main/java/org/testcontainers/containers/GenericContainer.java]: Where did this import come from?

File: core/src/test/java/org/testcontainers/junit/GenericContainerRuleTest.java [hunk 3]
```diff
@@ -124,6 +126,13 @@ public static void setupContent() throws FileNotFoundException {
             .withExtraHost("somehost", "192.168.1.10")
             .withCommand("/bin/sh", "-c", "while true; do cat /etc/hosts | nc -l -p 80; done");
 
+    /**
+     * Redis with shared memory
+     */
+    @ClassRule
```
Comment (kiview) [inline core/src/test/java/org/testcontainers/junit/GenericContainerRuleTest.java]: I think using the `@ClassRule` for all containers in this test class is quite bad and we should instead start the containers in the test itself. It's not your fault of course and should not be part of this PR, just a heads up for the others.

File: core/src/main/java/org/testcontainers/containers/GenericContainer.java [hunk 4]
```diff
@@ -252,7 +261,14 @@ private void tryStart(Profiler profiler) {
 
             logger().info("Creating container for image: {}", dockerImageName);
             profiler.start("Create container");
-            CreateContainerCmd createCommand = dockerClient.createContainerCmd(dockerImageName);
+
+            HostConfig hostConfig = new HostConfig();
```
Comment (kiview) [inline core/src/main/java/org/testcontainers/containers/GenericContainer.java]: I imagine we will continue adding support for more host config parameters, so I would prefer the `HostConfig` initialization in a private method, like `buildHostConfig()`.

File: core/src/test/java/org/testcontainers/junit/GenericContainerRuleTest.java [hunk 5]
```diff
@@ -375,4 +384,13 @@ public void addExposedPortAfterWithExposedPortsTest() {
         assertTrue("withExposedPort should be exposed", redis.getExposedPorts().contains(REDIS_PORT));
         assertTrue("addExposedPort should be exposed", redis.getExposedPorts().contains(8987));
     }
+
+    @Test
+    public void sharedMemorySetTest() {
+        assertEquals("Shared memory is not set", redisWithSharedMemory.getShmSize(), 1024 * FileUtils.ONE_MB);
```
Comment (kiview) [inline core/src/test/java/org/testcontainers/junit/GenericContainerRuleTest.java]: I would remove this assert.

File: core/src/test/java/org/testcontainers/junit/GenericContainerRuleTest.java (hunk 3 above)
Comment (fzakaria) [reply core/src/test/java/org/testcontainers/junit/GenericContainerRuleTest.java]: Moved to a try-with-resources ; also made it busybox to lighten the container that starts up.

**Comment by fzakaria** [comment]
Thanks for feedback @kiview  -- I've addressed them with latest update.
- moved to busybox + try-with-resources in the test
- moved HostConfig to a `buildHostConfig`
- removed extra assert

File: core/src/main/java/org/testcontainers/containers/GenericContainer.java [hunk 6]
```diff
@@ -1157,6 +1181,16 @@ public SELF withCreateContainerCmdModifier(Consumer<CreateContainerCmd> modifier
         return self();
     }
 
+    /**
+     * Size of /dev/shm
+     * @param bytes The number of megabybytestes to assign the shared memory. Null to not set any.
```
Comment (kiview) [inline core/src/main/java/org/testcontainers/containers/GenericContainer.java]:
```suggestion
     * @param bytes The number of megabytes to assign the shared memory. Null to not set any.
```

Doesn't `null` mean we will use the Docker default of 64MB?

File: core/src/main/java/org/testcontainers/containers/GenericContainer.java [hunk 7]
```diff
@@ -1157,6 +1181,16 @@ public SELF withCreateContainerCmdModifier(Consumer<CreateContainerCmd> modifier
         return self();
     }
 
+    /**
+     * Size of /dev/shm
+     * @param bytes The number of megabytes to assign the shared memory. If null, it will apply the Docker default which is 64 KB.
```
Comment (kiview) [inline core/src/main/java/org/testcontainers/containers/GenericContainer.java]:
```suggestion
     * @param bytes The number of megabytes to assign the shared memory. If null, it will apply the Docker default which is 64 MB.
```

Isn't it MB?

Comment (fzakaria) [reply core/src/main/java/org/testcontainers/containers/GenericContainer.java]: Darn it! I went off the Oracle docs but you are write.

http://www.tin.org/bin/man.cgi?section=8&topic=dockerd

```
    --default-shm-size=64MiB
         Set the daemon-wide default shm size for containers. Default is
       64MiB.
```

**Comment by fzakaria** [comment]
@kiview What's next step to get this merged & published?

**Comment by kiview** [comment]
Hey @fzakaria, sorry for letting this hanging, I think it's good to go.
I'll ping the others about final feedback and then we can merge.

File: core/src/main/java/org/testcontainers/containers/GenericContainer.java [hunk 8]
```diff
@@ -252,7 +260,12 @@ private void tryStart(Profiler profiler) {
 
             logger().info("Creating container for image: {}", dockerImageName);
             profiler.start("Create container");
-            CreateContainerCmd createCommand = dockerClient.createContainerCmd(dockerImageName);
+
+            HostConfig hostConfig = buildHostConfig();
```
Comment (bsideup) [inline core/src/main/java/org/testcontainers/containers/GenericContainer.java]: please move it to `applyConfiguration`

**Comment by fzakaria** [comment]
@bsideup made changes requested.

**Comment by fzakaria** [comment]
Update on this ? Quite a lot of effort to get a helpful change in.
I might abandon the process soon.

**Comment by fzakaria** [comment]
@kiview thanks. 
Sorry for the attitude -- other stuff that day was likely getting to me.

**Comment by kiview** [comment]
Travis failure unrelated, restarted.

**Comment by kiview** [comment]
@fzakaria merged, thanks a lot for the great PR as well as your patience :slightly_smiling_face:
